test/v1/controllers/RetrievePeriodicObligationsControllerSpec.scala

```diff
@@ -0,0 +1,184 @@
+
+        status(result) shouldBe OK
+        contentAsJson(result) shouldBe responseBody
+        header("X-CorrelationId", result) shouldBe Some(correlationId)
+      }
+    }
+
+    "return the error as per spec" when {
+      "parser errors occur" must {
+        def errorsFromParserTester(error: MtdError, expectedStatus: Int): Unit = {
+          s"a ${error.code} error is returned from the parser" in new Test {
+
+            MockRetrievePeriodicObligationsRequestParser
+              .parse(rawData)
+              .returns(Left(ErrorWrapper(Some(correlationId), error, None)))
+
+            val result: Future[Result] = controller.handleRequest(nino, Some(typeOfBusiness), Some(incomeSourceId), Some(fromDate), Some(toDate), Some(status))(fakeRequest)
+
+            status(result) shouldBe expectedStatus
+            contentAsJson(result) shouldBe Json.toJson(error)
+            header("X-CorrelationId", result) shouldBe Some(correlationId)
+          }
+        }
+
+        val input = Seq(
+          (NinoFormatError, BAD_REQUEST),
+          (TypeOfBusinessFormatError, BAD_REQUEST),
+          (BusinessIdFormatError, BAD_REQUEST),
+          (FromDateFormatError, BAD_REQUEST),
+          (ToDateFormatError, BAD_REQUEST),
+          (StatusFormatError, BAD_REQUEST),
+          (MissingFromDateError, BAD_REQUEST),
+          (MissingToDateError, BAD_REQUEST),
+          (ToDateBeforeFromDateError, BAD_REQUEST),
+          (MissingTypeOfBusinessError, BAD_REQUEST),
+          (RuleDateRangeInvalidError, BAD_REQUEST),
+          (RuleFromDateNotSupportedError, BAD_REQUEST)
+        )
+
+        input.foreach(args => (errorsFromParserTester _).tupled(args))
+      }
+
+    "service errors occur" must {
+      def serviceErrors(mtdError: MtdError, expectedStatus: Int): Unit = {
+        s"a $mtdError error is returned from the service" in new Test {
+
+          MockRetrievePeriodicObligationsRequestParser
+            .parse(rawData)
+            .returns(Right(requestData))
+
+          MockRetrievePeriodicObligationsService
+            .retrieve(requestData)
+            .returns(Future.successful(Left(ErrorWrapper(Some(correlationId), mtdError))))
+
+          val result: Future[Result] = controller.handleRequest(nino, Some(typeOfBusiness), Some(incomeSourceId), Some(fromDate), Some(toDate), Some(status))(fakeRequest)
+
+          status(result) shouldBe expectedStatus
+          contentAsJson(result) shouldBe Json.toJson(mtdError)
+          header("X-CorrelationId", result) shouldBe Some(correlationId)
+        }
+      }
+
+      val input = Seq(
+        (NinoFormatError, BAD_REQUEST),
+        (NotFoundError, NOT_FOUND),
+        (DownstreamError, INTERNAL_SERVER_ERROR),
+        (NoObligationsFoundError, NOT_FOUND)
+      )
+
+      input.foreach(args => (serviceErrors _).tupled(args))
+    }
+  }
+  }
```

Comment:
Indentation has gone wonky somewhere, these shouldn't be indented at the same level

Reply:
^ this again